### PR TITLE
Some unit tests for the cli

### PR DESCRIPTION
**Pull request by whyboris**
I'm happy to add more unit tests if they are welcome 👍

### CODE REVIEW

**Comment by rtfpessoa** [comment]
Overall I would love to have more unit and integration tests, maybe even some screnshot comparison to understand if the CSS is breaking something.

Regarding this examples specifically:
* I would like to keep the testing libraries to only one, since we already have mocha we should keep it or update the remaining tests if other library seems more suitable.
* The testing library should only be available as a DevDependency (should not be needed for runtime)
* This project was how I learned JS so the code is not very good (maybe even quite bad :disappointed:) for testing. In my daily work I really try to avoid mocks and rewrite the code to be more injectable so we can test most of it just by passing the needed parts.

In sum I have two questions:
* What is your experience with `sinon`? (I don't do JS daily to I am not following the trends/innovations) Do you think switching would be interesting?
* About the tests what do you think about using Dependency Injection / Method Passing instead of mocks for the majority of the tests?

**Comment by whyboris** [comment]
_Mocha_ [does not come equipped with spies](https://github.com/mochajs/mocha/wiki/Spies) and the library recommends [Sinon](https://github.com/cjohansen/sinon.js).

Spies are particularly useful for _unit_ testing. If the code is modular (and especially if it's _functional_) it's enough to test that function calls other functions / methods with appropriate arguments. For that we need to create a spy / stub.

Sorry I was tired and accidentally stuck _sinon_ in _Dependencies_ not _DevDependencies_ 😓 

At work we use [Jasmine](https://github.com/jasmine/jasmine) which includes spies and better logging when expected value is different from actual. It's easy to work with and given how few written tests there are, it would likely be simple to switch over.

I am more eager to get #42 to a state that's worth merging; we could revisit this PR after then 👍

**Comment by whyboris** [comment]
I've barely ever used _Sinon_ - just a bit here: https://github.com/whyboris/karma-helpful-reporter
I've not done much research so I'm unsure about the quality of anything other than _Jasmine_

I'm sorry I'm unsure what you mean with "Dependency Injection / Method passing" with respect to tests. Do you mean just to use the actual `Diff2HtmlInterface` and see if all the pieces work together by "expecting" correct output given standardized input?

**Comment by rtfpessoa** [comment]
@whyboris my idea was going more in this direction:
* https://hernantz.github.io/mock-yourself-not-your-tests.html
* https://philippe.bourgau.net/avoid-mocks-and-test-your-core-domain-faster-with-hexagonal-architecture/#
* https://www.artificialworlds.net/blog/2014/04/11/avoid-mocks-by-refactoring-to-functional/

but maybe this is not that important in this small application. If you feel like contributing some tests I will be ok accepting the mocks. Also if you think jasmine would be better feel free to change. I have nothing special regarding mockito.

**Comment by rtfpessoa** [comment]
@whyboris if you change the dependency to devDependencies I can merge this one.

**Comment by whyboris** [comment]
I'll look into failing tests tomorrow 👌 unsure what happened yet. 

Thank you for the links -- good reading material! 

I might try hooking up _Jasmine_ just to see how it goes 👍 

_edit_: the failing tests were because we now have the `ignore` argument in the `getInput` method. Fixed with https://github.com/rtfpessoa/diff2html-cli/pull/45/commits/cdabb617141921c340105e8f0d78815d383372f9

**Comment by whyboris** [comment]
Sorry for so many commits, please consider `Squash and merge` rather than `merge commit` whenever this PR is ready 😅

**Comment by whyboris** [comment]
Currently we're using [Istanbul](https://www.npmjs.com/package/istanbul) which is no longer maintained and [NYC](https://www.npmjs.com/package/nyc) is recommended instead.

Simply replacing `istanbul` with `nyc` in _package.json_ and switching the `coverage` script to `nyc mocha` we get this:

image

Would you like me to make the switch in this PR ? :)

**Comment by whyboris** [comment]
This is what the change for _NYC_ would look like: https://github.com/whyboris/diff2html-cli/commit/bfbee97c76a84fb6ced5044f4d7e550299d2bff6

It should keep all the functionality the same. As a bonus it will include a separate command for generating an HTML output `npm run coverage-html` 👍

**Comment by rtfpessoa** [comment]
> Would you like me to make the switch in this PR ? :)

Go ahead 👍

**Comment by whyboris** [comment]
_node 5_ fails with:
```
> nyc mocha

/root/diff2html-cli/node_modules/nyc/node_modules/find-up/index.js:5
module.exports = (filename, opts = {}) => {
                                 ^

SyntaxError: Unexpected token =
```
Will not work with `nyc` I suspect

_node 4_ fails with:
```
The package ajv@6.8.1 does not satisfy its siblings' peerDependencies requirements!
```
Fails during `npm install`
`ajv` seems to be part of `eslint` - perhaps other dependencies have it as a subdependency, and an early version of npm can't handle the different version complexity? Unsure why it passed before but not now otherwise.

**Comment by whyboris** [comment]
Now both _node4_ and _node5_ fail with the `Unexpected token =` in the _nyc_ library.

It looks like #44 PR will likely require abandoning earlier node versions anyway; do you think it would be acceptable / appropriate to call the next version a major release, dropping support for all node versions prior to 8?

The _Node_ [timeline](https://github.com/nodejs/Release) suggests that everyone should be at least on version 8 by April 2019 (node 6 leaves _maintenance_ status then).

**Comment by rtfpessoa** [comment]
@whyboris sure. That sounds appropriate.

**Comment by rtfpessoa** [comment]
We could also add 11 to the test list.

**Comment by whyboris** [comment]
Great 👍 -- please let me know if there's anything I should do with this PR for you to merge it.

Of course I don't mean to hurry you - no rush on any of this 😁

**Comment by rtfpessoa** [comment]
Remove the old node versions and add 11. That should be enough.

**Comment by whyboris** [comment]
😅 I hadn't realized that the configuration for this was in the .yml file 👍

File: .circleci/config.yml [hunk 1]
```diff
@@ -68,14 +60,18 @@ jobs:
     docker:
       - image: node:10
 
+  build-node_11:
+    <<: *latest-build
```
Comment (rtfpessoa) [inline .circleci/config.yml:64]: Sorry. Can you change `latest-build` to `common-build` in `build-node_10` line 59 so we don't push coverage twice?

**Review by rtfpessoa — CHANGES_REQUESTED** [review]

File: .circleci/config.yml (hunk 1 above)
Comment (whyboris) [reply .circleci/config.yml:64]: Sorry have never dealt with this file / config before -- should have been less cavalier when copy-pasting the last command (node 10 -> 11) 😓

**Pull request merged**